### DISCUSSION TITLE
Inconsistent Visualisation?

### DISCUSSION THREAD

**Discussion by idkwid2022** [pytorch-ignite/code-generator#219] [Q&A] [answered]
Hi,

I'm using image segmentation template, and modified the exp_logger so it use the trainer rather than the evaluator.

I'm doing this because I'm a total newbie on this subject and just begin to learn to use pytorch and everything, so I want to see which image is being trained, and stuff. 

I'm aware it will slowdown the system, and everything, but I need it to see so I can understand it better.

https://github.com/pytorch-ignite/code-generator/blob/f1372194000fb461c9e3eb7f4f85f43ad8f61a6c/src/templates/template-vision-segmentation/main.py#L130-L141

But I found the images on tensorboard kind of strange. Please look at the last column.

![image](https://user-images.githubusercontent.com/111631518/187151333-6c9f5d14-8b9f-4efb-94eb-616b8c539b52.png)

It looks like it's inverted?
The original image and mask are

| 1st Col| 3rd Col | 4th Col |
|----|----|----|
|![17](https://user-images.githubusercontent.com/111631518/187152215-c20c4f07-2e59-43bf-986b-5fe58f8fc594.jpg) ![17](https://user-images.githubusercontent.com/111631518/187152275-bcff22c4-92c0-40c6-867a-5a4202705e98.png) | ![20](https://user-images.githubusercontent.com/111631518/187152395-f718e659-f312-41a5-b451-f314929951d0.jpg) ![20](https://user-images.githubusercontent.com/111631518/187152449-09e7f344-fd48-44c8-94b5-b83a6f8db904.png)| ![26](https://user-images.githubusercontent.com/111631518/187151564-cd0612f6-e158-4c61-93b4-d73c2fd7b8b1.jpg) ![26](https://user-images.githubusercontent.com/111631518/187151634-cfff1c1f-0755-4578-897f-1dea92758185.png)|

So my question is, is this by design? I mean like dropout or something?

Thanks,

**Comment by vfdev-5** [comment]
Visualization should create a grid described here: https://github.com/pytorch-ignite/code-generator/blob/f1372194000fb461c9e3eb7f4f85f43ad8f61a6c/src/templates/template-vision-segmentation/vis.py#L50-L58


~~Looks like your 4th column image has ground truth like that. where ground truth mask is inverted. Can you check that explicitly in your data ?~~

> The original image and mask are

@idkwid2022 why do you have GT green masks for col 1 and 3 and red for col 4 ?

**Reply by idkwid2022** [reply]
@vfdev-5 they're supposed to be different classes, 2 classes, 3 with background.

These are my images and masks. It's just for learning, so nothing fancy, precision and all other is out of question. I just found it kind of strange, because the other images and masks are correctly shown, at least to me they are. 

The first image on my question is the result shown on tensorboard, so the visualization on vis.py are working, the grid and all, but somehow on some images, it comes out different.

![img-all](https://user-images.githubusercontent.com/111631518/187174562-c57da0cf-f057-48f3-abcc-dd23c0a7ae1c.png)

![mask-all](https://user-images.githubusercontent.com/111631518/187174640-1bdab5f3-93a5-4e71-8937-82cd74402da0.png)

This is another sample, please see the first and second column, it's image number 19 and 27, respectively from the dataset. The 27 is shown correctly, but 19 is not. Also you can see that the multiple class image is shown correctly on the 3rd column.

![image](https://user-images.githubusercontent.com/111631518/187178080-60cfe049-1f08-4ccd-952b-766b1e5cd428.png)

**Reply by vfdev-5** [reply]
I see. Can you try to visualize the mask of "4th Col" image (where you have background and class 1 inversion) using `render_mask` method: https://github.com/pytorch-ignite/code-generator/blob/f1372194000fb461c9e3eb7f4f85f43ad8f61a6c/src/templates/template-vision-segmentation/vis.py#L34-L39 and see if it is still correct ?

Next, you may make a batch of images with "4th Col" image and run `make_grid` to see if everything is rendered as expected.

 > The first image on my question is the result shown on tensorboard, so the visualization on vis.py are working, the grid and all, but somehow on some images, it comes out different.
 
So, you can confirm that `make_grid` and `render_mask` methods from `vis.py` are working as expected ?

**Reply by idkwid2022** [reply]
Alright, I'll try that, I'll get back later with the result. Thank you.

**Reply by vfdev-5** [reply]
@idkwid2022 another question, is your dataset public, such I could take a look from my side ?

**Reply by idkwid2022** [reply]
They should be, I just made them up from images from google, so they're not mine. I can send them to you if you want.

**Reply by vfdev-5** [reply]
OK, if you can put them on a shared drive such that I could download them it could be helpful.

**Reply by idkwid2022** [reply]
I just upload them to github: https://github.com/idkwid2022/learning-machine-learning

Fyi, I made it by cutting images from google with image editor, and create the mask manually using QuPath.

**Reply by vfdev-5** [reply]
Thanks, let me check them and the code from my side

**Reply by idkwid2022** [reply]
It looks like I'm gonna have to need more time to check the method one by one. But I just test with 'suspected images' by removing the other images and mask. Please see on the next comment (I think better to show it on another comment for clarity).

> I see. Can you try to visualize the mask of "4th Col" image (where you have background and class 1 inversion) using `render_mask` method:
> 
> https://github.com/pytorch-ignite/code-generator/blob/f1372194000fb461c9e3eb7f4f85f43ad8f61a6c/src/templates/template-vision-segmentation/vis.py#L34-L39
> 
> and see if it is still correct ?
> Next, you may make a batch of images with "4th Col" image and run `make_grid` to see if everything is rendered as expected.
> 
> > The first image on my question is the result shown on tensorboard, so the visualization on vis.py are working, the grid and all, but somehow on some images, it comes out different.
> 
> So, you can confirm that `make_grid` and `render_mask` methods from `vis.py` are working as expected ?

**Comment by idkwid2022** [comment]
I suspected that the images are part of the problem, so I remove all other images except some.

And I found image 6, 12, 14, 19, 21, 26, 38 are all inverted:

![image](https://user-images.githubusercontent.com/111631518/187188890-37551e4d-137f-4340-8c1d-42b43e58bdd6.png)
![image](https://user-images.githubusercontent.com/111631518/187189091-1af4dc58-9bff-40a6-a1eb-f9e728d287ce.png)

On tensorboard (These are only two grids for sample, the others are also inverted):


---
![image](https://user-images.githubusercontent.com/111631518/187189479-cf051bdf-ca95-4180-b6c8-710ca7383379.png)

---
![image](https://user-images.githubusercontent.com/111631518/187190431-8a838280-0ec2-42f4-8980-84364f2c9f4b.png)


I don't understand this. The images were all made by the same process, and all the same formats.

**Comment by idkwid2022** [comment, marked as answer]
Nvm, could you please forget this. It's my mistake.

I added `.quantize(256, method=Image.Quantize.FASTOCTREE)` on loading the mask on data.py.

I think it was because I want to show the image while debugging the code, and forgot to remove it.

~~~
    def __getitem__(self, index):
        img_from_file = Image.open(self.images[index]).convert("RGB")
        img = np.asarray(img_from_file)
        assert img is not None, f"Image at '{self.images[index]}' has a problem"
        mask_from_file = Image.open(self.masks[index]).quantize(256, method=Image.Quantize.FASTOCTREE)
        mask = np.asarray(mask_from_file)
        assert mask is not None, f"Mask at '{self.masks[index]}' has a problem"
~~~

I'm sorry.

Thank you for everything.

**Reply by vfdev-5** [reply]
OK, sounds good that you could find the issue.
From my side, visualization was correct:
![image](https://user-images.githubusercontent.com/2459423/187195511-3f05fbfa-b38e-4d4a-b50a-e6c750b6b536.png)

**Reply by idkwid2022** [reply]
Yes, of course. Now I can see the training progress with the image grids. 
It's fun to see it learning, confused and getting better.
Again, thank you.
